globus_automate_client/models.py

```diff
@@ -0,0 +1,392 @@
+    # Not supported by Flows
+    # OutputPath: t.Optional[str] = None
+
+    class Config:
+        extra = Extra.forbid
+
+    _enforce_jsonpath = validator(
+        "InputPath", "Result", "ResultPath", allow_reuse=True
+    )(enforce_jsonpath)
+    _check_jsonpath_syntax_in_dict = validator("Parameters", allow_reuse=True)(
+        check_jsonpath_syntax_in_dict
+    )
+
+
+class WaitState(NextOrEndState):
+    Type: te.Literal[StateType.Wait]
+    Comment: t.Optional[str] = None
+    InputPath: t.Optional[str] = None
+    OutputPath: t.Optional[str] = None
+    Seconds: t.Optional[int] = None
+    Timestamp: t.Optional[datetime.datetime] = None
+    SecondsPath: t.Optional[str] = None
+    TimestampPath: t.Optional[str] = None
+
+    _enforce_jsonpath = validator(
+        "InputPath", "OutputPath", "SecondsPath", "TimestampPath", allow_reuse=True
+    )(enforce_jsonpath)
+
+    class Config:
+        extra = Extra.forbid
+
+    @root_validator(skip_on_failure=True)
+    def validate_mutually_exclusive_fields(cls, values):
+        fields = ["Seconds", "Timestamp", "SecondsPath", "TimestampPath"]
+        set_fields = {f: values.get(f) for f in fields if values.get(f) is not None}
+        if len(set_fields) > 1:
+            raise ValueError(
+                f"Only one of these fields may be set: {set_fields.keys()}"
+            )
+        if len(set_fields) == 0:
+            raise ValueError(f"At least one field must be set: {fields}")
+
+        return values
+
+
+FlowState = te.Annotated[
+    t.Union[
+        PassState, WaitState, ActionState, FailState, ChoiceState, ExpressionEvalState
+    ],
+    Field(discriminator="Type"),
+]
+
+
+class FlowDefinition(BaseModel):
+    StartAt: str
+    States: t.Dict[str, FlowState]
+    Comment: t.Optional[str]
+
+    @validator("States")
+    def validate_states(cls, v):
+        if len(v) < 1:
+            raise ValueError('At least one "State" is required')
+
+        invalid_state_names = [
+            state_name
+            for state_name in v.keys()
+            if len(state_name) > 128 or len(state_name) == 0
+        ]
+        if invalid_state_names:
+            raise ValueError(
+                f"The following states have invalid names: {', '.join(invalid_state_names)}"
+            )
+        return v
+
+    @root_validator(skip_on_failure=True)
+    def validate_start_at_is_a_state(cls, values):
+        states, start_at = values.get("States"), values.get("StartAt")
+        if states and start_at not in states:
+            raise ValueError(f'StartAt state "{start_at}" is not defined in States')
+        return values
+
+    @root_validator(skip_on_failure=True)
+    def validate_all_transtion_states_exist(cls, values):
+        states = values.get("States", {})
+        defined_states = set(states.keys())
+        referenced_states = {values["StartAt"]}
+
+        for state in states.values():
+            if state.Type not in {StateType.Choice, StateType.Fail}:
```

Comment:
In my prototype, I used a mixin, abstract base class on the various `State` model classes to have them return the referenced state(s) (e.g. the value for `Next`). I think that reads rather cleanly as it puts that logic close to the state definition itself.

Reply:
That's a really interesting take on the problem, definitely an alternate solution!